### PR TITLE
[ISSUE-745][FEATURE] Change ShuffleClient to scala code

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### Related issues.
https://github.com/alibaba/celeborn/issues/745

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client/src/main/scala/org/apache/celeborn/client/ShuffleClient.scala [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.FileSystem
+
+import org.apache.celeborn.client.ShuffleClient._
+import org.apache.celeborn.client.read.RssInputStream
+import org.apache.celeborn.common.RssConf
+import org.apache.celeborn.common.protocol.message.ControlMessages._
+import org.apache.celeborn.common.rpc.RpcEndpointRef
+
+abstract class ShuffleClient extends Cloneable {
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/ShuffleClient.scala]: > Why should it extend cloneable? It is a singleton instance.

Copy from java code, removed

**Review by FMX — APPROVED** [review]
LGTM. Thanks.

File: client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala [hunk 2]
```diff
@@ -0,0 +1,1121 @@
...
+        Compressor.getCompressor(conf)
+      }
+    }
+
+  // key: shuffleId
+  private val reduceFileGroupsMap: Map[Integer, ReduceFileGroups] =
+    new ConcurrentHashMap[Integer, ReduceFileGroups]
+
+  def this(conf: RssConf, userIdentifier: UserIdentifier) {
+    this()
```
Comment (waitinfuture) [inline client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala:108]: Is this necessary?

File: client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala [hunk 3]
```diff
@@ -0,0 +1,1121 @@
...
+        loc.getEpoch,
+        loc,
+        cause)) {
+      callback.onFailure(new IOException("Revive Failed"))
+    } else {
+      if (mapperEnded(shuffleId, mapId, attemptId)) {
+        logDebug(s"Retrying push data, but the mapper(map $mapId attempt $attemptId) has ended.")
+        pushState.inFlightBatches.remove(batchId)
+      } else {
+        val newLoc: PartitionLocation = reducePartitionMap.get(shuffleId).get(partitionId)
```
Comment (waitinfuture) [inline client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala:157]: I think the type can be removed to be more concise, multiple places.

**Comment by waitinfuture** [comment]
This PR is critical, plz run extensive e2e test to check stability and correctness @FMX .

File: client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala:108]: > Is this necessary?

Yea

File: client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/ShuffleClientImpl.scala:157]: > I think the type can be removed to be more concise, multiple places.

Sure

**Pull request closed without merging**